### PR TITLE
compiler: introduce ZonGen and make `ast-check` run it for ZON inputs

### PR DESCRIPTION
**Pull request by mlugg**
Currently, `zig ast-check` fails on ZON files, because it tries to interpret the file as Zig source code. This commit introduces a new verification pass, `std.zig.ZonGen`, which applies to an AST in ZON mode.

Like `AstGen`, this pass also converts the AST into a more helpful format. Rather than a sequence of instructions like `Zir`, the output format of `ZonGen` is a new datastructure called `Zoir`. This type is essentially a simpler form of AST, containing only the information required for consumers of ZON. It is also far more compact than `std.zig.Ast`, with the size generally being comparable to the size of the well-formatted source file.

The emitted `Zoir` is currently not used aside from the `-t` option to `ast-check` which causes it to be dumped to stdout. However, in future, it can be used for comptime `@import` of ZON files, as well as for simpler handling of files like `build.zig.zon`, and even by other parts of the Zig Standard Library.

Resolves: #22078

---

cc @MasonRemaley

### CODE REVIEW

**Comment by MasonRemaley** [comment]
Some context for other readers--we discussed offline how this interacts with #20271.

The TLDR is that I'll update my PR to operate on ZonAst instead of Ast once this PR is merged. In doing so:
* I'll be able to deduplicate logic that currently needs to exist in both the runtime and comptime Zon lowering code
* I'll get to sidestep one of the major remaining TODOs on that PR, which is to implement friendly error messages for AST nodes that aren't valid ZON (e.g. types)

File: lib/std/zig/Zoir.zig [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+
+            pub fn at(r: Range, i: u32) Index {
+                assert(i < r.len);
+                return @enumFromInt(@intFromEnum(r.start) + i);
+            }
+        };
+    };
+};
+
+pub const NodeRepr = struct {
```
Comment (andrewrk) [inline lib/std/zig/Zoir.zig:138]: Suggest to move this inside `Node` and rename to `Repr`

Comment (mlugg) [reply lib/std/zig/Zoir.zig:138]: Good point, I actually meant to do that -- I'm signing off for tonight but will do this tomorrow (and look at any other review comments you leave)

File: lib/std/zig/ZonGen.zig [hunk 2]
```diff
@@ -0,0 +1,835 @@
...
+}
+
+fn lowerNumberError(zg: *ZonGen, err: std.zig.number_literal.Error, token: Ast.TokenIndex, bytes: []const u8) Allocator.Error!void {
+    const is_float = std.mem.indexOfScalar(u8, bytes, '.') != null;
+    switch (err) {
+        .leading_zero => if (is_float) {
+            try zg.addErrorTok(token, "number '{s}' has leading zero", .{bytes});
+        } else {
+            try zg.addErrorTokNotes(token, "number '{s}' has leading zero", .{bytes}, &.{
+                try zg.errNoteTok(token, "use '0o' prefix for octal ltierals", .{}),
```
Comment (Vexu) [inline lib/std/zig/ZonGen.zig:693]:
```suggestion
                try zg.errNoteTok(token, "use '0o' prefix for octal literals", .{}),
```

**Comment by Atomk** [comment]
In a couple doc comments in `Zoir.zig` there's `'token == 'invalid_token'` (with backticks) which will render incorrectly

**Pull request merged**